File: src/java/magma/core/stream/SingleHead.java
package magma.core.stream;

import magma.core.option.None;
import magma.core.option.Option;
import magma.core.option.Some;

public class SingleHead<T> implements Head<T> {
    private final T value;
    private boolean retrieved = false;

    public SingleHead(T value) {
        this.value = value;
    }

    @Override
    public Option<T> next() {
        if (retrieved) return new None<>();
        retrieved = true;
        return new Some<>(value);
    }
}

File: src/c/magma/stream/SingleHead.c
import magma.option.None;
import magma.option.Option;
import magma.option.Some;
struct SingleHead<T> implements Head<T> {private final T value;private boolean retrieved;public SingleHead(T value) {
        this.value = value;
        this.retrieved = false;
    }
	Option<T> next(){
		if (this.retrieved) {
			return None<>();
		}= true
		return Some<>(this.value);
	}
}